Issue 172: December Final Notice

ecedmondson opened the pull request:
December final notice

ecedmondson commented on tests/test_models.py:
```diff
@@ -139,3 +140,21 @@ def test_address_model_str_representation(address):
     )
+
+
+@pytest.mark.django_db
```
This test is probably worth careful attention.

ecedmondson commented:
Bike ride thoughts for self later: can't assume that the December 2023 recipients fit criteria of updated query. That will need to be a post 01/12/2023 change

reginafcompton commented on tests/test_management.py:
```diff
@@ -40,64 +56,158 @@ def mock_today_is_first_in_command():
+            )
+            for m in members
+        ]
+    )
```
The util function mocks work fine, though I wonder if you considered using a package to mock datetime instead, e.g., https://github.com/adamchainz/time-machine. (The AHS project uses `freezegun`...but we should stop, because it can be very slow.)

reginafcompton replied:
...anyway, it's absolutely not worth changing. these tests do what they need to do!

ecedmondson replied:
I don't normally like mocking datetime, as datetime usually mocks back, but I will give it a try if that's what you prefer.

ecedmondson replied:
Oh, I see. Time machine is not installed.

ecedmondson replied:
https://github.com/americanhandelsociety/americanhandelsociety-members/pull/180

reginafcompton commented on tests/test_models.py:
```diff
@@ -139,3 +140,21 @@ def test_address_model_str_representation(address):
+        [m.membership_type != Member.MembershipType.MESSIAH_CIRCLE for m in members]
+    ), f"Messiah Circle members cannot be overdue by definition."
+    assert all(
+        [m.date_of_last_membership_payment.year < year for m in members]
```
Should this be `m.date_of_last_membership_payment.year == year - 1`? (i.e., should more precisely assert on this filter: https://github.com/americanhandelsociety/americanhandelsociety-members/pull/179/files#diff-e716bbec71188cbd2a4cff67612148d45b0922d41c0d933ee315b398c99cc149R48 ?)

reginafcompton reviewed (commented):
Looks great! I left one comment that might be worth considering.

ecedmondson commented on americanhandelsociety_app/management/commands/send_overdue_payment_email.py:
```diff
@@ -84,9 +113,15 @@ class Command(BaseCommand):
         if today_is_first_of_month():
-            members_with_overdue_payments = get_members_with_overdue_payments()
-            sent_count = send_and_log(members_with_overdue_payments)
+            # Tech-Debt for Issue 172: the final notice for the first year of automated e-mails
```
@reginafcompton ...this may not be necessary. But I cannot assume that the members who have been receiving overdue payments this year have a payment year from 2022 (which is what the other filter would enforce). So I am continuing to use the existing filter provided by the function. After December first we can do a 1-2 line PR to move to the new queryset. How does that sound? Seem okay?

ecedmondson replied:
Oh, and then, I still left the QS changes in this PR because I didn't realize the above until I was on a bike ride. What do you think? Do you prefer I remove it from this PR?

reginafcompton replied:
> But I cannot assume that the members who have been receiving overdue payments this year have a payment year from 2022

oh, you know: I can just check....

****

all right, the five members who have been receiving the late notice all made payments in 2022. so, you can use `objects.dues_payment_pending` here.

ecedmondson replied:
Oh I didn't think to ask you. I'll drop that commit then

ecedmondson replied:
Thank you for checking!

reginafcompton replied:
Thank you for being so thoughtful!

ecedmondson commented:
I did some rebasing to get time machine in here, updated the tests, rebased the most recent commit to switch back to the qs method. I am testing from the shell too, but wouldn't mind other eyes.

reginafcompton commented on americanhandelsociety_app/models.py:
```diff
@@ -32,6 +33,20 @@ def create_superuser(self, email, password, **fields):
+        return self.exclude(
+            models.Q(is_member_via_other_organization=True)
+            | models.Q(membership_type=self.model.MembershipType.MESSIAH_CIRCLE)
+        ).filter(date_of_last_membership_payment__year=year_now() - 1)
```
This looks correct!

reginafcompton reviewed (approved):
This looks great. I did not manually test, however. (...but the changes look straightforward, so I am not worried.)

ecedmondson commented:
> This looks great. I did not manually test, however. (...but the changes look straightforward, so I am not worried.)

I think you are more confident than I am. That's fine!

reginafcompton commented:
I ran through the critical flows on the staging site, and nothing seems broken (which means the small upgrades did not introduce regressions).

I am going to deploy to production. Thanks again! (And happy thanksgiving.)

ecedmondson commented:
> I ran through the critical flows on the staging site, and nothing seems broken (which means the small upgrades did not introduce regressions).
> 
> 
> 
> I am going to deploy to production. Thanks again! (And happy thanksgiving.)

Thanks for letting me know! I'm so glad it seems to be working.

I was thinking about picking up another issue, if you don't mind. 

(And happy thanksgiving to you too!)
